ch_pipeline/analysis/source_removal.py

```diff
@@ -920,13 +1080,149 @@ def process(self, data, model, beams):
         for pp, upp in enumerate(model.index_map["pol"]):
 
             this_pol = np.flatnonzero(pol == upp)
+            dist_pol = distance[:, this_pol]
 
-            mdl = np.sum(
-                coeff[:, pp, np.newaxis, np.newaxis, :]
-                * source_model[:, this_pol, :, :],
-                axis=-1,
-            )
+            # Loop over frequencies
+            for ff, nu in enumerate(freq):
 
-            vis[:, this_pol, :] -= mdl
+                # Calculate source model
+                source_model, sedge = model_extended_sources(
+                    nu, dist_pol, timestamp, bodies, **source_model_kwargs
+                )
+                source_model = source_model[0]
+
+                # Multipy source model by the effective beam
+                for ss, src in enumerate(sources):
+                    this_beam = beams[src].vis[ff][this_pol].view(np.ndarray) * (
+                        beams[src].weight[ff][this_pol].view(np.ndarray) > 0.0
+                    ).astype(np.float32)
+                    source_model[..., sedge[ss] : sedge[ss + 1]] *= this_beam[
+                        ..., np.newaxis
+                    ]
+
+                mdl = np.sum(
+                    coeff[ff, pp, np.newaxis, np.newaxis, :] * source_model,
+                    axis=-1,
+                )
+
+                vis[ff, this_pol, :] -= mdl
 
         return data
+
+
+def kz_coeffs(m, k):
+    """Compute the coefficients for a Kolmogorov-Zurbenko filter.
+
+    Parameters
+    ----------
+    m : int
+        Size of the moving average window.
+    k : int
+        Number of iterations.
+
+    Returns
+    -------
+    coeff : np.ndarray
+        Array of size k * (m - 1) + 1 containing the filter coefficients.
+    """
+
+    # Coefficients at degree one
+    coef = np.ones(m, dtype=np.float64)
+
+    # Iterate k-1 times over coefficients
+    for i in range(1, k):
+
+        t = np.zeros((m, m + i * (m - 1)))
+        for km in range(m):
+            t[km, km : km + coef.size] = coef
+
+        coef = np.sum(t, axis=0)
+
+    assert coef.size == k * (m - 1) + 1
+
+    return coef / m ** k
+
+
+def apply_kz_lpf_2d(y, flag, window=3, niter=8, mode="wrap", frac_required=0.80):
+    """Apply a 2D Kolmogorov-Zurbenko filter.
```

Comment:
I appreciate I'm being a bit lazy, but it might be good to have a sentence about what a Kolmogorov-Zurbenko filter actually does for the ignorant (like myself).

Reply:
If you visit the Wikipedia page, you will be rewarded with some nice photos of Kolmogorov and Zurbenko themselves: https://en.wikipedia.org/wiki/Kolmogorov–Zurbenko_filter

Reply:
🤣 

That's hilarious. It's like their holiday photos or something, particularly the of them on the ship looking at handbags!